File: backend/timebank/src/main/java/kookmin/software/capstone2023/timebank/application/service/bank/transfer/TransferService.java
package kookmin.software.capstone2023.timebank.application.service.bank.transfer;

import kookmin.software.capstone2023.timebank.domain.model.BankAccountTransaction;
import lombok.Data;

import java.math.BigDecimal;

public interface TransferService {

    BankAccountTransaction transfer(TransferRequest request);

    @Data
    class TransferRequest {
        private final Long accountId;
        private final String senderAccountNumber;
        private final String receiverAccountNumber;
        private final BigDecimal amount;
        private final String password;

        public TransferRequest(Long accountId, String senderAccountNumber, String receiverAccountNumber, BigDecimal amount, String password) {
            this.accountId = accountId;
            this.senderAccountNumber = senderAccountNumber;
            this.receiverAccountNumber = receiverAccountNumber;
            this.amount = amount;
            this.password = password;
        }

        public TransferRequest(Long accountId, String senderAccountNumber, String receiverAccountNumber, BigDecimal amount) {
            this.accountId = accountId;
            this.senderAccountNumber = senderAccountNumber;
            this.receiverAccountNumber = receiverAccountNumber;
            this.amount = amount;
            this.password = null;
        }
    }
}

File: src/main/java/kookmin/software/capstone2023/timebank/application/service/bank/transfer/TransferService.kt
package kookmin.software.capstone2023.timebank.application.service.bank.transfer

import kookmin.software.capstone2023.timebank.domain.model.BankAccountTransaction
import java.math.BigDecimal

interface TransferService {
    fun transfer(request: TransferRequest): BankAccountTransaction
    data class TransferRequest(
        val accountId: Long,
        val senderAccountNumber: String,
        val receiverAccountNumber: String,
        val amount: BigDecimal,
        val password: String? = null,
    )
}
